scripts/bright/bm25qs_vs_bow_plots.py

```diff
@@ -0,0 +1,140 @@
+            "-m",
+            "ndcg_cut.10",
+            qrels,
+            f"runs/run.bright.bm25qs.{task}.txt",
+        ]
+        bm25qs_results = trec_eval(args, return_per_query_results=True)
+        args = [
+            "-c",
+            "-q",
+            "-m",
+            "ndcg_cut.10",
+            qrels,
+            f"runs/run.bright.bm25.{task}.txt",
+        ]
+        bm25_results = trec_eval(args, return_per_query_results=True)
+        for qid in query_lens.keys():
+            aggr_data[f"{task}_{qid}"] = (query_lens[qid], bm25qs_results[qid] - bm25_results[qid])
+    return aggr_data
+
+
+def unpack_xy(data):
+    """Return xs, ys as 1-D NumPy arrays from an iterable of (x, y)."""
+    xs, ys = zip(*data)  # built-in zip → tuples
+    return np.asarray(xs), np.asarray(ys)
+
+
+def scatter(xs, ys, log, output_path):
+    fig, ax = plt.subplots(figsize=(8, 4))
+    if log:
+        ax.set_xscale("log", base=2)
+    ax.scatter(xs, ys, s=15, alpha=0.7)
+    ax.set_xlabel("Query Length in Tokens (Log-Scale)", fontweight="bold")
+    ax.set_ylabel("Δ nDCG@10  (BM25 – BoW)", fontweight="bold")
+    ax.grid(True, which="both", alpha=0.3)
+    fig.tight_layout()
+    fig.savefig(output_path, bbox_inches="tight")
+    plt.close(fig)
+
+
+def violin_log_bins(
+    xs, ys, output_path, log_width=1, return_counts=True  # width in log2
+):  # debug helper
+    xs, ys = np.asarray(xs), np.asarray(ys)
+    if (xs < 1).any():
+        raise ValueError("All query lengths must be >= 1 for log-scale bins.")
+
+    # 1) bucket edges in log10 space
+    log_edges = np.arange(0, np.ceil(np.log2(xs.max())) + log_width, log_width)
+    edges = 2**log_edges
+
+    # 2) digitise and collect only non-empty groups
+    idx = np.digitize(xs, edges[1:])  # indices 0 … n-1
+    groups, labels, counts = [], [], []
+
+    for i in range(len(edges) - 1):
+        mask = idx == i
+        if mask.any():  # keep only **non-empty** bins
+            groups.append(ys[mask])
+            left = int(edges[i])
+            right = int(edges[i + 1] - 1)
+            labels.append(f"{left:,}-{right:,}")
+            counts.append(mask.sum())
+
+    # 3) plot – now safe because *all* groups have data
+    fig, ax = plt.subplots(figsize=(max(8, 0.65 * len(groups)), 4))
+    ax.violinplot(groups, showmeans=True, showextrema=True, showmedians=False)
+
+    ax.set_xticks(range(1, len(labels) + 1), labels, rotation=45, ha="right")
+    ax.axhline(0, color="grey", ls="--", lw=1)
+    ax.grid(axis="y", ls=":", alpha=0.3)
+
+    ax.set_xlabel(f"Log-Scale Bucket Widths", fontweight="bold")
+    ax.set_ylabel("Δ nDCG@10  (BM25 – BoW)", fontweight="bold")
+    fig.tight_layout()
+    fig.savefig(output_path, bbox_inches="tight")
+    plt.close(fig)
+
+    if return_counts:
+        return list(zip(labels, counts)), sum(counts)
+
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--output_dir", type=str, default="bm25vsbow_plots")
```

Comment:
House style - use `-` instead of `_` for improved readability - e.g., `--output-dir`